Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+This adapter class is generic over (1) any handler conforming to `LambdaHandler`, (2) the user-specified input and output types, and (3) any decoder and encoder conforming to `LambdaEventDecoder` and `LambdaOutputDecoder`. The adapter will wrap the underlying handler with encoding/decoding logic.
+
+#### Simplifying the handler protocols
+
+There are four different handler protocols in the current API (`SimpleLambdaHandler`, `LambdaHandler`, `EventLoopLambdaHandler`, `ByteBufferLambdaHandler`). As noted in the **Current Limitations** section, the ease-of-use varies for each handler protocol and users may not be able to easily determine which protocol best serves their use-case without spending time digging into the library. To reduce this problem and provide users with clear-cut options, we propose replacing all of the existing handler protocols with just two: `LambdaHandler` and `StreamingLambdaHandler`. Both will be explained in the **Detailed Solution** section.
+
+## Detailed Solution
+
+Below you’ll find explanation’s for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below), which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
+    /// Write a response part into the stream
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type `ByteBuffer`) will be removed.
+
+A new function `backgroundTask()` will also be added. This will allow tasks to be run in the background while and after the response is/has been sent. Please note that `LambdaContext` will not be Sendable anymore.
+
+```swift
+/// A context object passed as part of an invocation in LambdaHandler handle functions.
+public struct LambdaContext {
+    /// The request ID, which identifies the request that triggered the function invocation.
+    public var requestID: String { get }
+
+    /// The AWS X-Ray tracing header.
+    public var traceID: String { get }
+
+    /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
+    public var invokedFunctionARN: String { get }
+
+    /// The timestamp that the function times out.
+    public var deadline: DispatchWallTime { get }
```

Comment:
Can we use `ContinuousClock.Instant` here?

Reply:
Nope this is epoch time.

Reply:
As per [Lambda runtime API](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html#runtimes-api-next)

> Lambda-Runtime-Deadline-Ms – The date that the function times out in Unix time milliseconds. For example, 1542409706888.

Reply:
Got you so we would need a `UTCClock.Instant` here

Reply:
Yes. Sadly we don't have that yet. Only way out would be if Lambda built its own Clock. But I don't think we want to do that.

Reply:
Beware that [DispatchWallTime](https://developer.apple.com/documentation/dispatch/dispatchwalltime) has microsecond precision while the Lambda API uses milliseconds. The implementation will have to convert units

Reply:
@fabianfett one thing to be aware of though: There's no (documented) way to get the actual time out of a `DispatchWallTime`. The `.rawValue` is opaque and is going backwards, so _not_ UNIX time :)

```
  8> let wt = DispatchWallTime.now() 
wt: Dispatch.DispatchWallTime = {
  rawValue = 16723545949480398616
}
  9> let wt2 = DispatchWallTime.now()
wt2: Dispatch.DispatchWallTime = {
  rawValue = 16723545943513136616
}
 10> wt2 > wt
$R4: Bool = true
 11> wt2.rawValue > wt.rawValue
$R5: Bool = false
```

See how `wt2 > wt` as expected (I asked for `wt2` after `wt`) but  `wt2.rawValue > wt.rawValue == false`!!

So you can only really use `DispatchWallTime` if you use Dispatch itself to schedule events. You cannot convert it (in a documented way) to a timestamp or so.

Reply:
Maybe you should create a `LambdaUTCTimeClock` type that conforms to all the new hotness but implements the clock. With the hope that one day you can deprecate it and to `typealias LambdaUTCTimeClock = UTClock` with a bit of luck.

Reply:
> @weissi would you build the sleeping on top of dispatch?

No.

I'd just create a custom Clock type (because the language lacks `UTCClock` for some reason). Once you have your own clock (which you could for example back by `clock_gettime` and `struct timespec` but hidden as implementation detail).

Reply:
How would you implement the required sleeping methods though?
https://developer.apple.com/documentation/swift/clock/sleep(for:tolerance:)

Reply:
You can just use the normal `Task.sleep` on the non-suspending clock for that. UTC time and the non-suspending clock run at the same speed